### DISCUSSION TITLE
How to split one patch into multiple patches?

### DISCUSSION THREAD

**Discussion by patdx** [stacked-git/stgit#113] [Q&A] [answered]
I am pretty new to stacked-git, but it's really coming in handy for some code I am working on at the moment. 🙏 

However, I got a little stuck, I have a series of patches like this:

1. update package.json
2. update feature A
3. update package.json and feature B
4. update package.json and feature C

I want to squash all the package.json updates into the first patch, so the final stack of patches is like this:

1. update.package.json
2. update feature A
3. update feature B
4. update feature C

But I can't figure out a good way to split an existing patch?

I tried something like the `git rebase` approach [described here](https://git-scm.com/docs/git-rebase#_splitting_commits)

```bash
stg goto <patch-to-split>
git reset HEAD^ # rewind the index/staging area by 1 without changing the working tree
git add package.json
stg refresh --index --patch <first-patch>
```

However at this point I got an error:

```
Error: HEAD and top are not the same.
       This can happen if you modify a branch with git.
       "stg repair --help" explains more about what to do next.
stg refresh: Command aborted (all changes rolled back)
```

After I tried the `stg repair` command but it put me in the wrong patch.

Wonder if I am just thinking about this the wrong way? Would appreciate any advice or suggestions.

**Comment by patdx** [comment, marked as answer]
Hello, after I wrote this I started thinking about the problem again and started browsing through the source code. I found exactly what I want, the command is:

```
stg refresh --spill
```

> Additionally, the --spill option resets the topmost patch, emptying the patch while leaving the patch’s changes intact in the worktree.

https://stacked-git.github.io/man/stg-refresh/

So, for example, for my case to move the package.json update to another patch:

```bash
stg goto <patch-to-split>
stg refresh --spill
git reset # unstage all files, this seems to work fine?
git add package.json
stg refresh --index --patch <first-patch> # add "package.json" to the first patch instead
stg refresh # put the remaining files in the current patch
```

**Reply by topher200** [reply]
I even have an alias for this exact workflow.

```
sspl='stg refresh --spill && git reset .'
```

That will spill everything from the top patch into the worktree.

**Comment by habetsm-xilinx** [comment]
There are other ways to achieve what you want. It usually depends how larger/messy the commits are.
In your case you could have done:

1. Apply all patches
2. copy package.json somewhere.
3. stg goto patch-1
4. copy the new package.json
5. stg refresh
6. stg push, resolve conflicts.

**Comment by proski** [comment]
I wrote this script a while ago, I've been using it for years. I can submit it to the contrib directory is there is interest.
https://github.com/proski/development-tools/blob/master/stg-split/stg-split
